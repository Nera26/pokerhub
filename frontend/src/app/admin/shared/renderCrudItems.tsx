import type { Key, ReactNode } from 'react';

import type { AdminCrudItemsRenderProps } from '@/app/components/dashboard/common/AdminCrudPage';

type ItemContext<TIdentifier extends Key> = {
  submitting: boolean;
  deletingId: TIdentifier | null;
  identifier: TIdentifier;
  isDeleting: boolean;
};

export type CrudItemListProps<
  TItem,
  TIdentifier extends Key,
> = AdminCrudItemsRenderProps<TItem, TIdentifier> & {
  title: string;
  titleClassName?: string;
  loadingCopy: string;
  emptyCopy: string;
  getKey: (item: TItem) => TIdentifier;
  renderPrimary: (item: TItem) => ReactNode;
  description?: (item: TItem) => ReactNode;
  meta?: (item: TItem) => ReactNode;
  listClassName?: string;
  itemClassName?: string;
  contentClassName?: string;
  actionsClassName?: string;
  primaryClassName?: string;
  descriptionClassName?: string;
  metaClassName?: string;
  editButtonClassName?: string;
  deleteButtonClassName?: string;
  editLabel?: string;
  deleteLabel?: string;
  deletingLabel?: string;
  disableEdit?: (item: TItem, context: ItemContext<TIdentifier>) => boolean;
  disableDelete?: (item: TItem, context: ItemContext<TIdentifier>) => boolean;
  renderDeleteLabel?: (
    item: TItem,
    context: ItemContext<TIdentifier>,
  ) => ReactNode;
};

export function CrudItemList<TItem, TIdentifier extends Key>({
  title,
  titleClassName = 'text-lg font-semibold',
  loadingCopy,
  emptyCopy,
  getKey,
  renderPrimary,
  description,
  meta,
  items,
  loading,
  submitting,
  deletingId,
  startEdit,
  handleDelete,
  listClassName = 'space-y-2',
  itemClassName = 'flex items-center justify-between rounded border border-gray-300 px-3 py-2',
  contentClassName = 'flex flex-col',
  actionsClassName = 'flex items-center gap-3',
  primaryClassName = 'font-medium',
  descriptionClassName = 'text-sm text-gray-600',
  metaClassName = 'text-xs text-gray-500',
  editButtonClassName = 'text-blue-600 underline',
  deleteButtonClassName = 'text-red-600 underline',
  editLabel = 'Edit',
  deleteLabel = 'Delete',
  deletingLabel = 'Deletingâ€¦',
  disableEdit,
  disableDelete,
  renderDeleteLabel,
}: CrudItemListProps<TItem, TIdentifier>) {
  if (loading) {
    return (
      <section>
        <h2 className={titleClassName}>{title}</h2>
        <p>{loadingCopy}</p>
      </section>
    );
  }

  if (items.length === 0) {
    return (
      <section>
        <h2 className={titleClassName}>{title}</h2>
        <p>{emptyCopy}</p>
      </section>
    );
  }

  return (
    <section>
      <h2 className={titleClassName}>{title}</h2>
      <ul className={listClassName}>
        {items.map((item) => {
          const identifier = getKey(item);
          const itemContext: ItemContext<TIdentifier> = {
            submitting,
            deletingId,
            identifier,
            isDeleting: deletingId === identifier,
          };
          const descriptionContent = description?.(item);
          const metaContent = meta?.(item);
          const isEditDisabled = disableEdit
            ? disableEdit(item, itemContext)
            : submitting;
          const isDeleteDisabled = disableDelete
            ? disableDelete(item, itemContext)
            : submitting || itemContext.isDeleting;
          const deleteContent =
            renderDeleteLabel?.(item, itemContext) ??
            (itemContext.isDeleting ? deletingLabel : deleteLabel);

          return (
            <li key={identifier} className={itemClassName}>
              <div className={contentClassName}>
                <span className={primaryClassName}>{renderPrimary(item)}</span>
                {descriptionContent !== null &&
                descriptionContent !== undefined &&
                descriptionContent !== false ? (
                  <span className={descriptionClassName}>
                    {descriptionContent}
                  </span>
                ) : null}
                {metaContent !== null &&
                metaContent !== undefined &&
                metaContent !== false ? (
                  <span className={metaClassName}>{metaContent}</span>
                ) : null}
              </div>
              <div className={actionsClassName}>
                <button
                  type="button"
                  className={editButtonClassName}
                  onClick={() => startEdit(item)}
                  disabled={isEditDisabled}
                >
                  {editLabel}
                </button>
                <button
                  type="button"
                  className={deleteButtonClassName}
                  onClick={() => {
                    void handleDelete(identifier);
                  }}
                  disabled={isDeleteDisabled}
                >
                  {deleteContent}
                </button>
              </div>
            </li>
          );
        })}
      </ul>
    </section>
  );
}
