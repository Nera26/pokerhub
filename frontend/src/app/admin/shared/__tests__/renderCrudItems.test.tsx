import { render } from '@testing-library/react';

import { CrudItemList } from '../renderCrudItems';

describe('CrudItemList', () => {
  it('renders navigation items list', () => {
    const startEdit = jest.fn();
    const handleDelete = jest.fn(async () => {});
    const { container } = render(
      <CrudItemList
        title="Existing items"
        titleClassName="text-lg font-semibold mb-2"
        loadingCopy="Loading navigation items…"
        emptyCopy="No navigation items found."
        items={[
          {
            flag: 'home',
            href: '/',
            label: 'Home',
            order: 1,
            iconName: 'homeIcon',
          },
          {
            flag: 'support',
            href: '/support',
            label: 'Support',
            order: 2,
          },
        ]}
        loading={false}
        submitting={false}
        deletingId={null}
        startEdit={startEdit}
        handleDelete={handleDelete}
        getKey={(item) => item.flag}
        renderPrimary={(item) => (
          <>
            {item.label}{' '}
            <span className="text-xs text-gray-500">({item.flag})</span>
          </>
        )}
        description={(item) => `${item.href} · Order ${item.order}`}
        meta={(item) => (item.iconName ? <>Icon: {item.iconName}</> : null)}
        renderDeleteLabel={() => 'Delete'}
      />,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders admin tabs list with delete state', () => {
    const startEdit = jest.fn();
    const handleDelete = jest.fn(async () => {});
    const { container } = render(
      <CrudItemList
        title="Existing tabs"
        loadingCopy="Loading admin tabs…"
        emptyCopy="No runtime admin tabs found."
        items={[
          {
            id: 'reports',
            title: 'Reports',
            component: '@/reports',
            icon: 'faChart',
          },
          {
            id: 'audit',
            title: 'Audit',
            component: '@/audit',
            icon: undefined,
          },
        ]}
        loading={false}
        submitting={true}
        deletingId="audit"
        startEdit={startEdit}
        handleDelete={handleDelete}
        getKey={(item) => item.id}
        renderPrimary={(item) => item.title}
        primaryClassName="font-semibold"
        description={(item) => item.id}
        meta={(item) => (item.icon ? `Icon: ${item.icon}` : 'Missing icon')}
        itemClassName="flex items-center justify-between rounded border border-gray-200 px-3 py-2"
        disableEdit={(_, { submitting, isDeleting }) =>
          submitting || isDeleting
        }
        disableDelete={(_, { isDeleting }) => isDeleting}
      />,
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
