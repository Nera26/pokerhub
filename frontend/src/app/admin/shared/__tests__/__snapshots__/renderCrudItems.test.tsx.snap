// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CrudItemList renders admin tabs list with delete state 1`] = `
<section>
  <h2
    class="text-lg font-semibold"
  >
    Existing tabs
  </h2>
  <ul
    class="space-y-2"
  >
    <li
      class="flex items-center justify-between rounded border border-gray-200 px-3 py-2"
    >
      <div
        class="flex flex-col"
      >
        <span
          class="font-semibold"
        >
          Reports
        </span>
        <span
          class="text-sm text-gray-600"
        >
          reports
        </span>
        <span
          class="text-xs text-gray-500"
        >
          Icon: faChart
        </span>
      </div>
      <div
        class="flex items-center gap-3"
      >
        <button
          class="text-blue-600 underline"
          disabled=""
          type="button"
        >
          Edit
        </button>
        <button
          class="text-red-600 underline"
          type="button"
        >
          Delete
        </button>
      </div>
    </li>
    <li
      class="flex items-center justify-between rounded border border-gray-200 px-3 py-2"
    >
      <div
        class="flex flex-col"
      >
        <span
          class="font-semibold"
        >
          Audit
        </span>
        <span
          class="text-sm text-gray-600"
        >
          audit
        </span>
        <span
          class="text-xs text-gray-500"
        >
          Missing icon
        </span>
      </div>
      <div
        class="flex items-center gap-3"
      >
        <button
          class="text-blue-600 underline"
          disabled=""
          type="button"
        >
          Edit
        </button>
        <button
          class="text-red-600 underline"
          disabled=""
          type="button"
        >
          Deleting…
        </button>
      </div>
    </li>
  </ul>
</section>
`;

exports[`CrudItemList renders navigation items list 1`] = `
<section>
  <h2
    class="text-lg font-semibold mb-2"
  >
    Existing items
  </h2>
  <ul
    class="space-y-2"
  >
    <li
      class="flex items-center justify-between rounded border border-gray-300 px-3 py-2"
    >
      <div
        class="flex flex-col"
      >
        <span
          class="font-medium"
        >
          Home
           
          <span
            class="text-xs text-gray-500"
          >
            (
            home
            )
          </span>
        </span>
        <span
          class="text-sm text-gray-600"
        >
          / · Order 1
        </span>
        <span
          class="text-xs text-gray-500"
        >
          Icon: 
          homeIcon
        </span>
      </div>
      <div
        class="flex items-center gap-3"
      >
        <button
          class="text-blue-600 underline"
          type="button"
        >
          Edit
        </button>
        <button
          class="text-red-600 underline"
          type="button"
        >
          Delete
        </button>
      </div>
    </li>
    <li
      class="flex items-center justify-between rounded border border-gray-300 px-3 py-2"
    >
      <div
        class="flex flex-col"
      >
        <span
          class="font-medium"
        >
          Support
           
          <span
            class="text-xs text-gray-500"
          >
            (
            support
            )
          </span>
        </span>
        <span
          class="text-sm text-gray-600"
        >
          /support · Order 2
        </span>
      </div>
      <div
        class="flex items-center gap-3"
      >
        <button
          class="text-blue-600 underline"
          type="button"
        >
          Edit
        </button>
        <button
          class="text-red-600 underline"
          type="button"
        >
          Delete
        </button>
      </div>
    </li>
  </ul>
</section>
`;
